File: jssdk/src/services/AnalyticsService.ts
import { BaseService } from "@/services/BaseService";

/**
 * 事件数据接口（发送到服务端）
 */
export interface AnalyticsEvent {
    /** 事件类型（如 page_view, click, purchase） */
    event: string;
    /** 页面路径 */
    path?: string;
    /** 来源页面 */
    referrer?: string;
    /** 页面标题 */
    title?: string;
    /** 会话 ID（用于 UV 去重） */
    sessionId?: string;
    /** 页面加载耗时（毫秒） */
    perfMs?: number;
    /** 事件时间戳（毫秒） */
    timestamp?: number;
    /** 自定义属性 */
    props?: Record<string, any>;
}

/**
 * 用户属性接口
 */
export interface UserProps {
    userId?: string;
    [key: string]: any;
}

/**
 * Analytics SDK 配置接口
 */
export interface AnalyticsConfig {
    /** 是否自动采集 page_view 事件 */
    autoPageView?: boolean;
    /** 批量发送间隔（毫秒），默认 5000 */
    flushInterval?: number;
    /** 批量发送阈值（事件数量），默认 10 */
    flushThreshold?: number;
    /** 是否启用调试模式 */
    debug?: boolean;
    /** 自定义 sessionId（默认自动生成） */
    sessionId?: string;
}

/**
 * 服务端 Analytics 配置响应
 */
export interface ServerAnalyticsConfig {
    enabled: boolean;
    retention: number;
    flushInterval: number;
    hasS3: boolean;
}

/**
 * 统计数据响应
 */
export interface StatsResponse {
    summary: {
        totalPV: number;
        totalUV: number;
        bounceRate: number;
        avgDur: number;
    };
    daily: Array<{
        date: string;
        pv: number;
        uv: number;
    }>;
    startDate: string;
    endDate: string;
}

/**
 * Top Pages 响应
 */
export interface TopPagesResponse {
    pages: Array<{
        path: string;
        pv: number;
        visitors: number;
    }>;
    startDate: string;
    endDate: string;
}

/**
 * Top Sources 响应
 */
export interface TopSourcesResponse {
    sources: Array<{
        source: string;
        visitors: number;
    }>;
    startDate: string;
    endDate: string;
}

/**
 * Devices 响应
 */
export interface DevicesResponse {
    browsers: Array<{
        name: string;
        visitors: number;
    }>;
    os: Array<{
        name: string;
        visitors: number;
    }>;
    startDate: string;
    endDate: string;
}

/**
 * Raw Logs 响应
 */
export interface RawLogsResponse {
    dates: string[];
}

const STORAGE_KEY_OPT_OUT = "pb_analytics_opt_out";
const STORAGE_KEY_SESSION_ID = "pb_analytics_session_id";

/**
 * AnalyticsService 提供用户行为分析功能。
 *
 * 使用示例：
 * ```js
 * const pb = new PocketBase("http://127.0.0.1:8090");
 *
 * // 初始化 Analytics（自动采集 page_view）
 * await pb.analytics.init();
 *
 * // 手动埋点
 * pb.analytics.track("click_buy", { price: 99, productId: "123" });
 *
 * // 关联登录用户
 * pb.analytics.identify({ userId: "user123", plan: "pro" });
 *
 * // GDPR 合规：用户选择退出
 * pb.analytics.optOut();
 * 
 * // === 管理员 API（需要 Superuser 认证）===
 * // 获取统计数据
 * const stats = await pb.analytics.getStats("7d");
 * 
 * // 获取热门页面
 * const pages = await pb.analytics.getTopPages("7d", 10);
 * ```
 */
export class AnalyticsService extends BaseService {
    private eventQueue: AnalyticsEvent[] = [];
    private flushTimer: ReturnType<typeof setTimeout> | null = null;
    private config: AnalyticsConfig = {
        autoPageView: true,
        flushInterval: 5000,
        flushThreshold: 10,
        debug: false,
    };
    private userProps: UserProps = {};
    private isInitialized = false;
    private sessionId: string = "";
    private isEnabled: boolean | null = null;

    /**
     * 初始化 Analytics 服务。
     *
     * @param config 可选配置
     */
    async init(config?: AnalyticsConfig): Promise<void> {
        if (this.isInitialized) {
            return;
        }

        if (config) {
            this.config = { ...this.config, ...config };
        }

        // 检查是否已 opt-out
        if (this.isOptedOut()) {
            this.log("Analytics is opted out, skipping initialization");
            return;
        }

        // 检查后端是否启用 Analytics
        const enabled = await this.checkAnalyticsEnabled();
        if (!enabled) {
            this.log("Analytics is disabled on server, skipping initialization");
            return;
        }

        // 获取或生成 session ID
        this.sessionId = this.config.sessionId || this.getOrCreateSessionId();

        // 自动采集 page_view
        if (this.config.autoPageView && typeof window !== "undefined") {
            this.trackPageView();
            this.setupHistoryListener();
        }

        // 设置页面卸载时发送事件
        if (typeof window !== "undefined") {
            window.addEventListener("beforeunload", () => this.flush(true));
            window.addEventListener("pagehide", () => this.flush(true));
        }

        // 启动定时 flush
        this.startFlushTimer();

        this.isInitialized = true;
        this.log("Analytics initialized with sessionId:", this.sessionId);
    }

    /**
     * 跟踪自定义事件。
     *
     * @param event 事件名称
     * @param props 事件属性
     */
    track(event: string, props?: Record<string, any>): void {
        if (this.isOptedOut()) {
            return;
        }

        const analyticsEvent: AnalyticsEvent = {
            event,
            path: typeof window !== "undefined" ? window.location.pathname : "",
            referrer: typeof document !== "undefined" ? document.referrer : "",
            title: typeof document !== "undefined" ? document.title : "",
            sessionId: this.sessionId,
            props: props || {},
            timestamp: Date.now(),
        };

        this.eventQueue.push(analyticsEvent);
        this.log("Event tracked:", event, props);

        // 检查是否需要立即 flush
        if (this.eventQueue.length >= (this.config.flushThreshold || 10)) {
            this.flush();
        }
    }

    /**
     * 跟踪页面浏览事件。
     * @param perfMs 页面加载耗时（毫秒），可选
     */
    trackPageView(perfMs?: number): void {
        const props: Record<string, any> = {
            url: typeof window !== "undefined" ? window.location.href : "",
        };
        
        // 添加性能数据
        if (perfMs !== undefined) {
            props.perfMs = perfMs;
        } else if (typeof performance !== "undefined") {
            props.perfMs = Math.round(performance.now());
        }

        this.track("page_view", props);
    }

    /**
     * 关联用户属性。
     *
     * @param props 用户属性
     */
    identify(props: UserProps): void {
        if (this.isOptedOut()) {
            return;
        }

        this.userProps = { ...this.userProps, ...props };
        this.log("User identified:", props);

        // 发送 identify 事件
        this.track("identify", props);
    }

    /**
     * 用户选择退出分析（GDPR 合规）。
     */
    optOut(): void {
        if (typeof localStorage !== "undefined") {
            localStorage.setItem(STORAGE_KEY_OPT_OUT, "true");
        }
        this.eventQueue = [];
        this.stopFlushTimer();
        this.log("User opted out of analytics");
    }

    /**
     * 用户选择重新加入分析。
     */
    optIn(): void {
        if (typeof localStorage !== "undefined") {
            localStorage.removeItem(STORAGE_KEY_OPT_OUT);
        }
        this.init(this.config);
        this.log("User opted in to analytics");
    }

    /**
     * 检查用户是否已退出分析。
     */
    isOptedOut(): boolean {
        if (typeof localStorage === "undefined") {
            return false;
        }
        return localStorage.getItem(STORAGE_KEY_OPT_OUT) === "true";
    }

    /**
     * 立即发送所有缓存的事件。
     *
     * @param useBeacon 是否使用 Beacon API（页面卸载时使用）
     */
    async flush(useBeacon = false): Promise<void> {
        if (this.eventQueue.length === 0) {
            return;
        }

        const events = [...this.eventQueue];
        this.eventQueue = [];

        // 转换为服务端期望的格式
        const payload = {
            events: events.map(e => ({
                event: e.event,
                path: e.path,
                referrer: e.referrer,
                title: e.title,
                sessionId: e.sessionId || this.sessionId,
                perfMs: e.props?.perfMs,
                timestamp: e.timestamp,
                // 其他 props 展开
                ...e.props,
            })),
        };

        this.log("Flushing events:", events.length);

        try {
            if (useBeacon && typeof navigator !== "undefined" && navigator.sendBeacon) {
                // 使用 Beacon API（页面卸载时）
                const url = this.client.buildURL("/api/analytics/events");
                navigator.sendBeacon(url, JSON.stringify(payload));
            } else {
                // 使用 fetch
                await this.client.send("/api/analytics/events", {
                    method: "POST",
                    body: payload,
                });
            }
        } catch (error) {
            this.log("Failed to flush events:", error);
            // 失败时将事件放回队列
            this.eventQueue = [...events, ...this.eventQueue];
        }
    }

    // ==================== 管理员 API ====================

    /**
     * 获取统计数据概览。
     * 需要 Superuser 认证。
     *
     * @param range 日期范围（today, 7d, 30d, 90d）
     */
    async getStats(range: string = "7d"): Promise<StatsResponse> {
        return this.client.send("/api/analytics/stats", {
            method: "GET",
            query: { range },
        });
    }

    /**
     * 获取热门页面排行。
     * 需要 Superuser 认证。
     *
     * @param range 日期范围
     * @param limit 返回条数（默认 10，最大 100）
     */
    async getTopPages(range: string = "7d", limit: number = 10): Promise<TopPagesResponse> {
        return this.client.send("/api/analytics/top-pages", {
            method: "GET",
            query: { range, limit },
        });
    }

    /**
     * 获取流量来源排行。
     * 需要 Superuser 认证。
     *
     * @param range 日期范围
     * @param limit 返回条数（默认 10，最大 100）
     */
    async getTopSources(range: string = "7d", limit: number = 10): Promise<TopSourcesResponse> {
        return this.client.send("/api/analytics/top-sources", {
            method: "GET",
            query: { range, limit },
        });
    }

    /**
     * 获取设备/浏览器统计。
     * 需要 Superuser 认证。
     *
     * @param range 日期范围
     */
    async getDevices(range: string = "7d"): Promise<DevicesResponse> {
        return this.client.send("/api/analytics/devices", {
            method: "GET",
            query: { range },
        });
    }

    /**
     * 获取可下载的原始日志日期列表。
     * 需要 Superuser 认证。
     */
    async getRawLogs(): Promise<RawLogsResponse> {
        return this.client.send("/api/analytics/raw-logs", {
            method: "GET",
        });
    }

    /**
     * 获取原始日志下载 URL。
     *
     * @param date 日期（格式：YYYY-MM-DD）
     */
    getRawLogDownloadURL(date: string): string {
        return this.client.buildURL(`/api/analytics/raw-logs/${date}`);
    }

    /**
     * 获取服务端 Analytics 配置。
     * 需要 Superuser 认证。
     */
    async getServerConfig(): Promise<ServerAnalyticsConfig> {
        return this.client.send("/api/analytics/config", {
            method: "GET",
        });
    }

    // ==================== 私有方法 ====================

    /**
     * 获取或创建 session ID。
     */
    private getOrCreateSessionId(): string {
        if (typeof localStorage === "undefined") {
            return this.generateSessionId();
        }

        let sessionId = localStorage.getItem(STORAGE_KEY_SESSION_ID);
        if (!sessionId) {
            sessionId = this.generateSessionId();
            localStorage.setItem(STORAGE_KEY_SESSION_ID, sessionId);
        }
        return sessionId;
    }

    /**
     * 生成唯一的 session ID。
     */
    private generateSessionId(): string {
        // 优先使用 crypto.randomUUID
        if (typeof crypto !== "undefined" && crypto.randomUUID) {
            return crypto.randomUUID();
        }
        // 降级方案：时间戳 + 随机数
        const timestamp = Date.now().toString(36);
        const random = Math.random().toString(36).substring(2, 15);
        return `${timestamp}-${random}`;
    }

    /**
     * 设置 History API 监听器（SPA 路由变化）。
     */
    private setupHistoryListener(): void {
        if (typeof window === "undefined") {
            return;
        }

        // 监听 popstate 事件（浏览器前进/后退）
        window.addEventListener("popstate", () => {
            this.trackPageView();
        });

        // 拦截 pushState 和 replaceState
        const originalPushState = history.pushState;
        const originalReplaceState = history.replaceState;

        history.pushState = (...args) => {
            originalPushState.apply(history, args);
            this.trackPageView();
        };

        history.replaceState = (...args) => {
            originalReplaceState.apply(history, args);
            this.trackPageView();
        };
    }

    /**
     * 启动定时 flush。
     */
    private startFlushTimer(): void {
        if (this.flushTimer) {
            return;
        }

        this.flushTimer = setInterval(() => {
            this.flush();
        }, this.config.flushInterval || 5000);
    }

    /**
     * 停止定时 flush。
     */
    private stopFlushTimer(): void {
        if (this.flushTimer) {
            clearInterval(this.flushTimer);
            this.flushTimer = null;
        }
    }

    /**
     * 调试日志。
     */
    private log(...args: any[]): void {
        if (this.config.debug) {
            console.log("[PocketBase Analytics]", ...args);
        }
    }

    /**
     * 检查后端是否启用 Analytics。
     * 缓存结果以避免重复请求。
     */
    private async checkAnalyticsEnabled(): Promise<boolean> {
        // 如果已经检查过，返回缓存结果
        if (this.isEnabled !== null) {
            return this.isEnabled;
        }

        try {
            // 尝试调用 config 接口检查是否启用
            // 注意：config 接口需要 Superuser 认证，所以我们用另一种方式
            // 直接尝试发送一个空事件，如果返回 404 说明未启用
            await this.client.send("/api/analytics/events", {
                method: "POST",
                body: { events: [] },
            });
            // 如果没有抛出错误，说明 Analytics 是启用的（即使返回错误 "No events provided"）
            this.isEnabled = true;
            this.log("Analytics enabled on server");
            return true;
        } catch (error: any) {
            // 如果是 404，说明 Analytics 未启用
            if (error?.status === 404) {
                this.isEnabled = false;
                this.log("Analytics is disabled on server");
                return false;
            }
            // 如果是 400（No events provided），说明 Analytics 是启用的
            if (error?.status === 400) {
                this.isEnabled = true;
                this.log("Analytics enabled on server");
                return true;
            }
            // 其他错误，假设未启用
            this.log("Failed to check analytics status:", error);
            this.isEnabled = false;
            return false;
        }
    }

    /**
     * 获取 Analytics 是否在服务端启用。
     * 如果尚未检查，返回 null。
     */
    get serverEnabled(): boolean | null {
        return this.isEnabled;
    }

    /**
     * 获取当前 session ID。
     */
    get currentSessionId(): string {
        return this.sessionId;
    }
}

File: pythonsdk/src/pocketbase/services/analytics_service.py
"""AnalyticsService for PocketBase analytics API.

提供用户行为分析功能，包括：
- 事件追踪（track_event, track_events）
- 统计查询（get_stats, get_top_pages, get_top_sources, get_devices）
- 原始日志（get_raw_logs, get_raw_log_download_url）
- 配置查询（get_config）

示例用法：
```python
from pocketbase import PocketBase

pb = PocketBase("http://127.0.0.1:8090")

# 追踪事件
pb.analytics.track_event(
    event="page_view",
    path="/home",
    session_id="user-session-123",
)

# 批量追踪
pb.analytics.track_events([
    {"event": "page_view", "path": "/home", "sessionId": "sess-1"},
    {"event": "click", "path": "/buy", "sessionId": "sess-1"},
])

# 获取统计数据（需要 Superuser 认证）
pb.collection("_superusers").auth_with_password("admin@example.com", "password")
stats = pb.analytics.get_stats(range="7d")
print(f"Total PV: {stats['summary']['totalPV']}")
print(f"Total UV: {stats['summary']['totalUV']}")

# 获取热门页面
pages = pb.analytics.get_top_pages(range="7d", limit=10)
for page in pages["pages"]:
    print(f"{page['path']}: {page['pv']} PV, {page['visitors']} UV")
```
"""

from typing import Any, TypedDict


from pocketbase.services.base_service import BaseService


class StatsSummary(TypedDict):
    """统计汇总数据。"""
    totalPV: int
    totalUV: int
    bounceRate: float
    avgDur: float


class DailyStats(TypedDict):
    """每日统计数据。"""
    date: str
    pv: int
    uv: int


class StatsResponse(TypedDict):
    """统计数据响应。"""
    summary: StatsSummary
    daily: list[DailyStats]
    startDate: str
    endDate: str


class PageStats(TypedDict):
    """页面统计数据。"""
    path: str
    pv: int
    visitors: int


class TopPagesResponse(TypedDict):
    """热门页面响应。"""
    pages: list[PageStats]
    startDate: str
    endDate: str


class SourceStats(TypedDict):
    """来源统计数据。"""
    source: str
    visitors: int


class TopSourcesResponse(TypedDict):
    """流量来源响应。"""
    sources: list[SourceStats]
    startDate: str
    endDate: str


class DeviceStats(TypedDict):
    """设备统计数据。"""
    name: str
    visitors: int


class DevicesResponse(TypedDict):
    """设备统计响应。"""
    browsers: list[DeviceStats]
    os: list[DeviceStats]
    startDate: str
    endDate: str


class RawLogsResponse(TypedDict):
    """原始日志响应。"""
    dates: list[str]


class ServerConfig(TypedDict):
    """服务端配置响应。"""
    enabled: bool
    retention: int
    flushInterval: int
    hasS3: bool


class EventsResponse(TypedDict):
    """事件提交响应。"""
    accepted: int
    total: int


class AnalyticsService(BaseService):
    """Service for interacting with analytics endpoints.
    
    提供事件追踪和统计查询功能。
    
    事件追踪方法（无需认证）：
    - track_event: 追踪单个事件
    - track_events: 批量追踪事件
    
    统计查询方法（需要 Superuser 认证）：
    - get_stats: 获取统计概览
    - get_top_pages: 获取热门页面
    - get_top_sources: 获取流量来源
    - get_devices: 获取设备统计
    - get_raw_logs: 获取原始日志列表
    - get_config: 获取服务端配置
    """

    def track_event(
        self,
        event: str,
        path: str,
        session_id: str | None = None,
        referrer: str | None = None,
        title: str | None = None,
        perf_ms: int | None = None,
        timestamp: int | None = None,
        **extra: Any,
    ) -> EventsResponse:
        """追踪单个分析事件。
        
        Args:
            event: 事件类型（如 "page_view", "click", "purchase"）
            path: 页面路径
            session_id: 会话 ID（用于 UV 去重），强烈建议提供
            referrer: 来源页面 URL
            title: 页面标题
            perf_ms: 页面加载耗时（毫秒）
            timestamp: 事件时间戳（毫秒），默认使用当前时间
            **extra: 其他自定义属性
            
        Returns:
            EventsResponse: 包含 accepted 和 total 字段
            
        示例:
            >>> pb.analytics.track_event(
            ...     event="purchase",
            ...     path="/checkout/success",
            ...     session_id="user-123",
            ...     order_id="ORD-456",
            ...     amount=99.99,
            ... )
            {'accepted': 1, 'total': 1}
        """
        event_data: dict[str, Any] = {
            "event": event,
            "path": path,
        }
        
        if session_id:
            event_data["sessionId"] = session_id
        if referrer:
            event_data["referrer"] = referrer
        if title:
            event_data["title"] = title
        if perf_ms is not None:
            event_data["perfMs"] = perf_ms
        if timestamp is not None:
            event_data["timestamp"] = timestamp
        
        # 添加额外的自定义属性
        event_data.update(extra)

        return self.track_events([event_data])

    def track_events(self, events: list[dict[str, Any]]) -> EventsResponse:
        """批量追踪分析事件。
        
        Args:
            events: 事件数据列表，每个事件应包含：
                - event: 事件类型（必填）
                - path: 页面路径（必填）
                - sessionId: 会话 ID（强烈建议）
                - referrer: 来源页面
                - title: 页面标题
                - perfMs: 页面加载耗时
                - timestamp: 事件时间戳（毫秒）
                - 其他自定义字段
            
        Returns:
            EventsResponse: 包含 accepted（成功接收数）和 total（总数）
            
        示例:
            >>> pb.analytics.track_events([
            ...     {"event": "page_view", "path": "/home", "sessionId": "sess-1"},
            ...     {"event": "click", "path": "/buy", "sessionId": "sess-1"},
            ... ])
            {'accepted': 2, 'total': 2}
        """
        return self.client.send(
            "/api/analytics/events",
            method="POST",
            body={"events": events},
        )

    def get_stats(
        self,
        range: str = "7d",  # noqa: A002
    ) -> StatsResponse:
        """获取统计数据概览。
        
        需要 Superuser 认证。
        
        Args:
            range: 日期范围，支持：
                - "today": 今天
                - "7d": 最近 7 天（默认）
                - "30d": 最近 30 天
                - "90d": 最近 90 天
            
        Returns:
            StatsResponse: 包含 summary（汇总）和 daily（每日数据）
            
        示例:
            >>> stats = pb.analytics.get_stats("7d")
            >>> print(f"Total PV: {stats['summary']['totalPV']}")
            >>> print(f"Total UV: {stats['summary']['totalUV']}")
            >>> for day in stats['daily']:
            ...     print(f"{day['date']}: {day['pv']} PV")
        """
        return self.client.send(
            "/api/analytics/stats",
            method="GET",
            params={"range": range},
        )

    def get_top_pages(
        self,
        range: str = "7d",  # noqa: A002
        limit: int = 10,
    ) -> TopPagesResponse:
        """获取热门页面排行。
        
        需要 Superuser 认证。
        
        Args:
            range: 日期范围（today, 7d, 30d, 90d）
            limit: 返回条数（默认 10，最大 100）
            
        Returns:
            TopPagesResponse: 包含 pages 列表，每个页面有 path, pv, visitors
            
        示例:
            >>> pages = pb.analytics.get_top_pages("7d", limit=5)
            >>> for page in pages["pages"]:
            ...     print(f"{page['path']}: {page['pv']} PV, {page['visitors']} UV")
        """
        return self.client.send(
            "/api/analytics/top-pages",
            method="GET",
            params={"range": range, "limit": limit},
        )

    def get_top_sources(
        self,
        range: str = "7d",  # noqa: A002
        limit: int = 10,
    ) -> TopSourcesResponse:
        """获取流量来源排行。
        
        需要 Superuser 认证。
        
        Args:
            range: 日期范围（today, 7d, 30d, 90d）
            limit: 返回条数（默认 10，最大 100）
            
        Returns:
            TopSourcesResponse: 包含 sources 列表，每个来源有 source 和 visitors
            
        示例:
            >>> sources = pb.analytics.get_top_sources("30d")
            >>> for src in sources["sources"]:
            ...     print(f"{src['source']}: {src['visitors']} visitors")
        """
        return self.client.send(
            "/api/analytics/top-sources",
            method="GET",
            params={"range": range, "limit": limit},
        )

    def get_devices(
        self,
        range: str = "7d",  # noqa: A002
    ) -> DevicesResponse:
        """获取设备/浏览器统计。
        
        需要 Superuser 认证。
        
        Args:
            range: 日期范围（today, 7d, 30d, 90d）
            
        Returns:
            DevicesResponse: 包含 browsers 和 os 列表
            
        示例:
            >>> devices = pb.analytics.get_devices("7d")
            >>> print("浏览器分布:")
            >>> for browser in devices["browsers"]:
            ...     print(f"  {browser['name']}: {browser['visitors']}")
            >>> print("操作系统分布:")
            >>> for os in devices["os"]:
            ...     print(f"  {os['name']}: {os['visitors']}")
        """
        return self.client.send(
            "/api/analytics/devices",
            method="GET",
            params={"range": range},
        )

    def get_config(self) -> ServerConfig:
        """获取服务端 Analytics 配置。
        
        需要 Superuser 认证。
        
        Returns:
            ServerConfig: 包含 enabled, retention, flushInterval, hasS3
            
        示例:
            >>> config = pb.analytics.get_config()
            >>> print(f"Analytics enabled: {config['enabled']}")
            >>> print(f"Data retention: {config['retention']} days")
        """
        return self.client.send(
            "/api/analytics/config",
            method="GET",
        )

    def get_raw_logs(self) -> RawLogsResponse:
        """获取可下载的原始日志日期列表。
        
        需要 Superuser 认证。
        
        Returns:
            RawLogsResponse: 包含 dates 列表（YYYY-MM-DD 格式）
            
        示例:
            >>> logs = pb.analytics.get_raw_logs()
            >>> for date in logs["dates"]:
            ...     print(f"Available: {date}")
        """
        return self.client.send(
            "/api/analytics/raw-logs",
            method="GET",
        )

    def get_raw_log_download_url(self, date: str) -> str:
        """获取指定日期原始日志的下载 URL。
        
        Args:
            date: 日期（格式：YYYY-MM-DD）
            
        Returns:
            完整的下载 URL
            
        示例:
            >>> url = pb.analytics.get_raw_log_download_url("2024-01-15")
            >>> print(url)  # http://127.0.0.1:8090/api/analytics/raw-logs/2024-01-15
        """
        return self.client.build_url(f"/api/analytics/raw-logs/{date}")

    def is_enabled(self) -> bool:
        """检查服务端 Analytics 是否启用。
        
        通过发送空事件请求来检测。
        
        Returns:
            bool: True 如果启用，False 如果禁用
            
        示例:
            >>> if pb.analytics.is_enabled():
            ...     pb.analytics.track_event("page_view", "/home", "sess-1")
        """
        try:
            # 发送空事件请求
            result = self.client.send(
                "/api/analytics/events",
                method="POST",
                body={"events": []},
            )
            # 如果返回错误但不是 404，说明 Analytics 启用了
            return True
        except Exception as e:
            # 检查是否是 404 错误
            if hasattr(e, "status") and e.status == 404:  # type: ignore
                return False
            # 400 错误（No events provided）说明 Analytics 启用了
            if hasattr(e, "status") and e.status == 400:  # type: ignore
                return True
            # 其他错误，假设未启用
            return False
